extension/src/nmost.rs

```diff
@@ -0,0 +1,264 @@
+                        return Some(state);
+                    }
+                    state.new_entry(*val);
+                }
+
+                Some(state)
+            } else {
+                Internal::new::<NMostTransState<T>>((sorted_vals, capacity).into()).to_inner()
+            }
+        })
+    }
+}
+
+fn nmost_trans_combine<T: Clone + Ord + Copy>(
+    first: Option<Inner<NMostTransState<T>>>,
+    second: Option<Inner<NMostTransState<T>>>,
+) -> Option<Inner<NMostTransState<T>>> {
+    match (first, second) {
+        (None, None) => None,
+        (None, Some(only)) | (Some(only), None) => unsafe {
+            Internal::new(only.clone()).to_inner()
+        },
+        (Some(a), Some(b)) => {
+            let mut a = a.clone();
+            // This could be made more efficient by iterating in the appropriate order with an early exit, but would requiring ordering the other heap
+            for entry in b.heap.iter() {
+                a.new_entry(*entry);
+            }
+            unsafe { Internal::new(a).to_inner() }
+        }
+    }
+}
+
+// TODO: serialize and deserialize will need to be implemented with Datum handling code
+#[derive(Clone, Debug)]
+pub struct NMostByTransState<T: Ord> {
+    values: NMostTransState<(T, usize)>,
+    data: Vec<Datum>,
+    oid: pg_sys::Oid,
+}
+
+impl<T: Clone + Ord> NMostByTransState<T> {
+    fn new(capacity: usize, first_val: T, first_element: pgx::AnyElement) -> NMostByTransState<T> {
+        // first entry will always have index 0
+        let first_val = (first_val, 0);
+        NMostByTransState {
+            values: NMostTransState::new(capacity, first_val),
+            data: vec![unsafe { deep_copy_datum(first_element.datum(), first_element.oid()) }],
+            oid: first_element.oid(),
+        }
+    }
+
+    fn new_entry(&mut self, new_val: T, new_element: pgx::AnyElement) {
+        assert!(new_element.oid() == self.oid);
+        if self.data.len() < self.values.capacity {
+            // Not yet full, easy case
+            self.values.new_entry((new_val, self.data.len()));
+            self.data
+                .push(unsafe { deep_copy_datum(new_element.datum(), new_element.oid()) });
+        } else if self
+            .values
+            .belongs_in_heap(&(new_val.clone(), self.data.len()))
+        {
+            // Full and value belongs in the heap (using len() for this check just keeps us from
+            // succeeding if we tie the max heap element)
+
+            let (_, index_to_replace) = *self
+                .values
+                .heap
+                .peek()
+                .expect("Can't be empty in this case");
+            let _old_datum = std::mem::replace(&mut self.data[index_to_replace], unsafe {
+                deep_copy_datum(new_element.datum(), new_element.oid())
+            });
+            // TODO: should we cleaning up 'old_datum' here?  There don't seem to be any convenience functions for this...
```

Comment:
Is this a potential leak?  I don't see where Datum implements Drop...

Reply:
Well, everything here is done in the aggregate context, so the memory will be reclaimed with the aggregate finishes, which should be fine in almost every case.  However, I couldn't quite convince myself that we weren't at risk of running out of memory if we ran this on a very large dataset, sorted such that every new element replaces an old value.  So I added a free_datum function that I'm now using here.